python/ray/serve/_private/result_wrapper.py

```diff
@@ -0,0 +1,126 @@
+import threading
+import time
+from abc import ABC, abstractmethod
+from typing import Callable, Optional
+
+import ray
+from ray.serve._private.utils import calculate_remaining_timeout
+
+
+class ResultWrapper(ABC):
+    @abstractmethod
+    def get(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def get_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def next(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def anext(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def resolve_generator_to_ref_sync(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def resolve_generator_to_ref_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def resolve_as_top_level_arg(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def add_callback(self, callback: Callable):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def cancel(self):
+        raise NotImplementedError
+
+
+class ActorResultWrapper(ResultWrapper):
```

Comment:
Would be nice to add some unit test for each method and the expected behavior.

Reply:
hmm, looked into this but it's kind of difficult because most of the wrapper's logic is just interacting directly with the ray object, so it's inherently not very unit-testable ... instead, having a ResultWrapper helps to unit test other objects that use it, e.g. the handle or router. I think for now, e2e handle tests do a good job of testing this, I can continue to brainstorm whether we can guard this with other forms of tests?

Reply:
I think you can make fake ray object ref/ generator objects and implement the needed methods :) (but if it's too much of a work, I'm fine with this being a follow up as well)

Reply:
yeah, I understand you can do that, but I think if I were to do that, it's still hard to test functionality. A lot of the test code might end up being duplicate of the actual code (by checking empty methods of fake object refs are called). I'll think more about it too, there might be a small subset of unit tests that can actually be useful.